doc/developer/design/20250415_large_select_result_size.md

```diff
@@ -0,0 +1,174 @@
+# Large Select Result Size
+
+## Context
+
+Currently, results of SELECT queries are sent back from the cluster to
+`environmentd` via the compute protocol and these results are fully
+materialized (stored in memory) in `environmentd` before sending them out to
+the client.
+
+This has several implications:
+
+- Sending large results "clogs up" the cluster/controller communication
+- The amount of memory we want to give to `environmentd` limits the size of
+  results we can return
+
+In practice the above make it so we limit the size of results using a
+`max_result_size` parameter, and some larger customers are chafing against
+that.
+
+Today, we always materialize the results of a query in `environmentd` memory
+before sending them on to the client. This is only required for some types of
+queries, though: only queries that require _post processing_ need
+materialization while queries that don't need post processing could be streamed
+through to the client. Queries that need an ordering established on the whole
+result require post processing, which are queries that have at least one of
+ORDER BY, LIMIT, or OFFSET. This is handled by `RowSetFinishing` in the code.
+
+Below, we will use _streamable queries_ for queries that don't require post
+processing, and _non-streamable queries_ for those that do require it.
```

Comment:
We always imply implicit post processing, i.e., we sort the data even if the user didn't specify any specific order. How would this impact your design?

Specifically, we might write down as many blobs as we have workers on the cluster replica. Each is sorted within itself. When sending the data to the client, we need to stream all blobs and merge them on-the-fly. Certainly possible, but at what resource utilization?